### ISSUE TITLE
[BUG] Server list artifacts http 500 error for proxied uploads

### ISSUE THREAD

**Issue by joaquinco** [mlflow/mlflow#5377]
Thank you for submitting an issue. Please refer to our [issue policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md) for additional information about bug reports. For help with debugging your code, please refer to [Stack Overflow](https://stackoverflow.com/questions/tagged/mlflow).

**Please fill in this bug report template to ensure a timely and thorough response.**

### Willingness to contribute
The MLflow Community encourages bug fix contributions. Would you or another member of your organization be willing to contribute a fix for this bug to the MLflow code base?

- [ ] Yes. I can contribute a fix for this bug independently.
- [x] Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.
- [ ] No. I cannot contribute a bug fix at this time.

### System information
- **Have I written custom code (as opposed to using a stock example script provided in MLflow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: `python:3.9.9-slim-bullseye` docker image
- **MLflow installed from (source or binary)**: binary
- **MLflow version (run ``mlflow --version``)**: 1.23.0
- **Python version**: 3.9.9
- **npm version, if running the dev UI**:
- **Exact command to reproduce**: See below

### Describe the problem

I have a mlflow server setup which is working fine except from listing artifacts from the ui after having uploaded them.

### Code to reproduce issue

I'm running mlflow server as follows:

```bash
mlflow  server \
      --host 0.0.0.0 \
      --port 5000 \
      --workers 4 \
      --backend-store-uri postgresql://$DATABASE_USER:$DATABASE_PASSWORD@postgres:5432/mlflow?sslmode=disable \
      --artifacts-destination s3://my-bucket/artifacts \
      --serve-artifacts \
      --gunicorn-opts "--access-logfile -"
```

And my test experiment is:

```python
import random
import mlflow

mlflow.set_tracking_uri('http://localhost:5000')
mlflow.set_experiment('test-1')

def run_experiment():
    mlflow.log_param('name', 'Peter')
    mlflow.log_param('age', 20 + random.random() * 30)

    output_path = 'output.txt'
    with open(output_path, 'w') as f:
        f.write('Something {}'.format(random.randint(30, 50)))

    mlflow.log_artifact(output_path)


with mlflow.start_run():
    run_experiment()
```

### Other info / logs

Error screenshot:
(I'm getting error 500 on http://localhost:5000/ajax-api/2.0/preview/mlflow/artifacts/list?run_uuid=a86a0e638a1748d3a48587f2d1e00488)

![Screen Shot 2022-02-14 at 13 42 56](https://user-images.githubusercontent.com/20804419/153907513-d9c8bc86-aa89-4685-91cb-c44bc3cdb6c6.png)

Server log:

```
2022/02/14 16:44:25 ERROR mlflow.server: Exception on /ajax-api/2.0/preview/mlflow/artifacts/list [GET]
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 2073, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1518, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1516, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1502, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
  File "/usr/local/lib/python3.9/site-packages/mlflow/server/handlers.py", line 238, in wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/mlflow/server/handlers.py", line 298, in wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/mlflow/server/handlers.py", line 571, in _list_artifacts
    artifact_entities = _get_artifact_repo(run).list_artifacts(path)
AttributeError: 'Response' object has no attribute 'list_artifacts'
172.22.0.1 - - [14/Feb/2022:16:44:25 +0000] "GET /ajax-api/2.0/preview/mlflow/artifacts/list?run_uuid=a86a0e638a1748d3a48587f2d1e00488 HTTP/1.1" 500 290 "http://localhost:5000/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/97.0.4692.99 Safari/537.36"
```

### Fix idea

In the `mlflow/server/handlers.py` file, in the `_list_artifacts()` method, I've tried using `_get_artifact_repo_mlflow_artifacts()` to fetch the repository instead of trying to fetch it from the run with `_get_artifact_repo(run)` and seems to have worked. Note that the run artifact's schema is `mlflow-artifacts:/` which, to my point of view, does not make sense to be used from the server side.

### What component(s), interfaces, languages, and integrations does this bug affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [x] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [x] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

**Comment by BenWilson2** [comment]
@joaquinco thank you for bringing this to our attention! We'll be working on a fix for this.

**Comment by BradAF** [comment]
I can confirm updating from release 1.23.1 to release 1.24.0 fixed this issue in my environment.

**Comment by joaquinco** [comment]
Yeap, worked for me too. Thanks for your work!

**Comment by BenWilson2** [comment]
@joaquinco thank you for bringing it to our attention. We truly appreciate community feedback like this.

**Comment by leviplj** [comment]
Found a problem...

If I start the UI server first, using an empty database, and then start the artifact server, the artifacts are not shown on the UI. However, if I start the artifact server first, it works and the files are shown on the UI.

OBS: There are no errors on the log. Before starting the server, delete the database.

```bash
# UI Server
mlflow server --host 0.0.0.0 \
  --port 8000 \
  --serve-artifacts \
  --backend-store-uri sqlite:///./site.db \
  --default-artifact-root mlflow-artifacts://localhost:8001
```
```bash
# Artifact Server
export AWS_ACCESS_KEY_ID="AWS_KID"
export AWS_SECRET_ACCESS_KEY="AWS_SKEY"
mlflow server --host 0.0.0.0 \
  --port 8001 \
  --serve-artifacts \
  --artifacts-only \
  --backend-store-uri sqlite:///./site.db \
  --artifacts-destination s3://mlflow-test-bucket
```

```python
# main.py
import mlflow
import os

mlflow.set_tracking_uri('http://localhost:8000')
with mlflow.start_run():
    mlflow.log_param("a", 1)
    mlflow.log_metric("b", 2)

    if not os.path.exists("outputs"):
        os.makedirs("outputs")

    with open("outputs/test.txt", "w") as f:
        f.write("hello world")

    print('log artifacts s')
    mlflow.log_artifact("outputs/test.txt")
    mlflow.set_tracking_uri('http://localhost:8000')
    print('log artifacts e')

```

**Comment by leviplj** [comment]
I noticed that when starting the ui first, the URI for the experiment on the table is `mlflow-artifacts://0.0.0.0:8001/0`, and when starting the artifact server first the URI is `mlflow-artifacts://0`

**Comment by iavinas** [comment]
These bugs are fixed in 1.24.0, (Yesterday's release)

**Comment by leviplj** [comment]
I've tested it today (01/03) around 13:00 UTC.

**Comment by dbczumar** [comment]
@leviplj The issue seems to be that your `UI Server` *and* your `Artifact Server` are both configured with `--serve-artifacts`. When you attempt to list artifacts via the UI, a ListArtifacts request is sent to the `UI Server`, which thinks that it's serving artifacts and attempts to resolve them from the default `--artifacts-destination` of `./mlruns`. If you restart your `UI Server` and omit `--serve-artifacts`, everything should work as intended.

We'll work on adding a warning here and clarifying the docs.

**Comment by leviplj** [comment]
@dbczumar Not using `--serve-artifacts` on the `UI Server`, throws the following error on the terminal when I open the run on the browser:
```bash
2022/03/02 11:00:07 ERROR mlflow.server: Exception on /ajax-api/2.0/preview/mlflow/artifacts/list [GET]
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 2073, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 1518, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 1516, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 1502, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
  File "/usr/local/lib/python3.8/site-packages/mlflow/server/handlers.py", line 327, in wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/mlflow/server/handlers.py", line 385, in wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/mlflow/server/handlers.py", line 676, in _list_artifacts
    artifact_entities = _get_artifact_repo(run).list_artifacts(path)
AttributeError: 'Response' object has no attribute 'list_artifacts'
```

And even without the `--serve-artifacts` on the `UI Server`, if I start it before the `Artifacts Server`, the experiment is saved with the URI `mlflow-artifacts://0.0.0.0:8001/0`.
Notice that I always create an empty database before starting the servers.

**Comment by BenWilson2** [comment]
Hi @leviplj I was able to repro your issue with the following configs:

Artifacts server:
```
mlflow server --host 0.0.0.0 \
              --port 5500 \
              --artifacts-destination /tmp/mlflow-testing \
              --backend-store-uri sqlite:///./mlflow-backend-store.db \
              --gunicorn-opts "--log-level debug" \
              --serve-artifacts \
              --artifacts-only
```
Tracking server:
```
mlflow server --host 0.0.0.0 \
              --port 5000 \
              --backend-store-uri sqlite:///./mlflow-backend-store.db \
              --default-artifact-root http://localhost:5500/api/2.0/mlflow-artifacts/artifacts/myruns \
              --gunicorn-opts "--log-level debug"
```
However, by omitting the `backend-store-uri` option from the artifacts server configuration (which is not needed for proxied operation since it is not handling run information, experiment tracking, metrics logging, etc and only handles artifact saving and loading) this should (hopefully) fix your problem. 

Please validate these configs for the servers to ensure that your environment is functioning correctly:

Artifacts server:
```
mlflow server --host 0.0.0.0 \
              --port 5500 \
              --artifacts-destination /tmp/mlflow-testing \
              --gunicorn-opts "--log-level debug" \
              --serve-artifacts \
              --artifacts-only
```
Tracking server:
```
mlflow server --host 0.0.0.0 \
              --port 5000 \
              --backend-store-uri sqlite:///./mlflow-backend-store.db \
              --default-artifact-root http://localhost:5500/api/2.0/mlflow-artifacts/artifacts/myruns \
              --gunicorn-opts "--log-level debug"
```

**Comment by BenWilson2** [comment]
We're going to be implementing controls for the cli to ensure that these incompatible configurations won't raise these sorts of frustrating problems for server administrators. Stay tuned!

**Comment by leviplj** [comment]
@BenWilson2 that works, but it's not using the new schema `mlflow-artifacts:/`, but `http`. Using `mlflow-artifacts` still breaks the tracking server.
If I don't set `--serve-artifacts`, the artifacts fail to return with the message `Unable to list artifacts stored under <code>{artifactUri}</code> for the current run ...`
If I set `--serve-artifacts`, no artifact is listed but there's no error message.

I guess the main problem is why the servers are behaving differently depending on which server I start first when using `--backend-store-uri` on both servers.
